### PR TITLE
feat: Add OpenDAL resource support

### PR DESCRIPTION
**Pull request by Xuanwo**
## Description of change

Close https://github.com/shuttle-hq/shuttle/issues/1565

This PR will add opendal sources support that users can connect 50+ storage services including s3, azblob, gcs, oss, .....

## How has this been tested? (if applicable)

unit test.

### CODE REVIEW

File: deployer/src/persistence/resource/mod.rs [hunk 1]
```diff
@@ -81,6 +81,7 @@ pub enum Type {
     Persist,
     Turso,
     Metadata,
+    Opendal,
```
Comment (Xuanwo) [inline deployer/src/persistence/resource/mod.rs:84]: I'm not sure if `Opendal` is a good name, maybe we can use `Blob` or `Storage` to better describe what it does?

Comment (jonaro00) [reply deployer/src/persistence/resource/mod.rs:84]: It would be better to not add a variant here and instead use `Custom`. The reason for this is that we are trying to use this enum for resources that Shuttle can manage, and all client plugins are separate.

File: resources/opendal/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+) -> ShuttleAxum {}
+```
+
+### Parameters
+
+| Parameter | Type  | Default    | Description                                      |
+|-----------|-------|------------|--------------------------------------------------|
+| scheme    | `str` | `"memory"` | The scheme of the storage service to connect to. |
+
+All secrets are loaded from your `Secret.toml` file. Visit the [OpenDAL Documentation](https://opendal.apache.org/docs/rust/opendal/services/index.html) for more information on how to setup the secrets for the storage service you want to connect to.
```
Comment (jonaro00) [inline resources/opendal/README.md:37]: An example of this would be helpful, such as "Example: if you are using S3, you need to set the secrets AAA, BBB, and CCC. Visit the docs for more details: ..."

**Review by jonaro00 — COMMENTED** [review]
Looks like a great addition!
We have some major refactorings of the resource code planned, and I think we'll at least want https://github.com/shuttle-hq/shuttle/pull/1543 done before adding this. I'll let you know of the progress.
Some small comments:

File: resources/opendal/README.md (hunk 2 above)
Comment (Xuanwo) [reply resources/opendal/README.md]: Added.

**Comment by Xuanwo** [comment]
Happy to see #1543 has been merge! Are there any work I need to do for this PR?

**Review by jonaro00 — COMMENTED** [review]
It looks good to me. I will check with the team for further approvals.

File: resources/opendal/README.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+### Parameters
+
+| Parameter | Type  | Default    | Description                                      |
+|-----------|-------|------------|--------------------------------------------------|
+| scheme    | `str` | `"memory"` | The scheme of the storage service to connect to. |
+
+All secrets are loaded from your `Secret.toml` file. 
+
+For instance, when using `s3`, you can configure the scheme to `s3` and specify the secrets: `bucket`, `access_key_id`, and `secret_access_key`.
```
Comment (oddgrd) [inline resources/opendal/README.md:39]: This is a clever way to load the config vars! However, there might be more secrets in `Secrets.toml` that are not related to the OpenDAL resource. Will that be a problem, or will they simply be ignored when initializing the operator?

**Review by oddgrd — COMMENTED** [review]
This is cool, thank you! I left one question. Other than that, an example in https://github.com/shuttle-hq/shuttle-examples would be terrific, but if you don't have time, we can also get to it. :slightly_smiling_face:

File: resources/opendal/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    fn from(error: Error) -> Self {
+        let msg = format!("Failed to build opendal resource: {:?}", error.0);
+        ShuttleError::Custom(CustomError::msg(msg))
+    }
+}
+
+#[async_trait]
+impl ResourceBuilder for Opendal {
+    const TYPE: Type = Type::Custom;
+    type Config = Self;
```
Comment (jonaro00) [inline resources/opendal/src/lib.rs:50]: For this type of plugin it is better to use `()` as the config type (see some of the other plugins), since there is nothing that the shuttle servers should know about this resource.

File: resources/opendal/README.md (hunk 3 above)
Comment (jonaro00) [reply resources/opendal/README.md:39]: Looks like that's the case.

File: resources/opendal/README.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+Users can connect OpenDAL to access data from a variety of storage services, including: s3, azblob, gcs, oss and [so on](https://opendal.apache.org/docs/rust/opendal/services/index.html).
+
+## Usage
+
+**IMPORTANT**: Currently Shuttle isn't able to provision a storage for you (yet). This means you will have to create the storage service first and setup the secrets accordingly.
+
+Add `shuttle-opendal` to the dependencies for your service by running `cargo add shuttle-opendal`.
+This resource will be provided by adding the `shuttle_opendal::Opendal` attribute to your Shuttle `main` decorated function.
+
+It returns a `opendal::OpenDAL` for you to connect the storage service.
```
[the remark this answers is not in the document]
Comment (Xuanwo) [reply resources/opendal/README.md:14]: Fixed.

File: resources/opendal/src/lib.rs (hunk 4 above)
Comment (Xuanwo) [reply resources/opendal/src/lib.rs]: Fixed.

**Comment by Xuanwo** [comment]
> Other than that, an example in [shuttle-hq/shuttle-examples](https://github.com/shuttle-hq/shuttle-examples) would be terrific, but if you don't have time, we can also get to it. 🙂

I'm willing to add examples! If I understand correctly, we need to release opendal resource first?

**Review by jonaro00 — APPROVED** [review]
